cpu/sam0_common/periph/flashpage.c

```diff
@@ -45,6 +42,53 @@ void flashpage_write(int page, void *data)
         PAC1->WPCLR.reg = NVMCTRL_PAC_BIT;
     }
 #endif
+}
+
+static void _lock(void)
+{
+    /* put peripheral access lock for the NVMCTRL peripheral */
+#ifdef CPU_FAM_SAML21
+    PAC->WRCTRL.reg = (PAC_WRCTRL_KEY_SET | ID_NVMCTRL);
+#else
+    if (PAC1->WPCLR.reg & NVMCTRL_PAC_BIT) {
+        PAC1->WPSET.reg = NVMCTRL_PAC_BIT;
+    }
+#endif
+}
+
+void flashpage_write_raw(void *target_addr, void *data, size_t len)
+{
+    /* The actual minimal block size for writing is 16B, thus we
+     * assert we write on multiples and no less of that length.
+     */
+    assert(!(len % FLASHPAGE_RAW_BLOCKSIZE));
+
+    /* ensure 4 byte aligned writes */
+    assert(!(((unsigned)target_addr % FLASHPAGE_RAW_ALIGNMENT) ||
+            ((unsigned)data % FLASHPAGE_RAW_ALIGNMENT)));
+
```

Comment:
One thing I just noticed (sorry, did not see this before): We do currently not at all check, if the given actually fits into the flash, or if we would overflow at the end of the flash. So say we have 8Kb flash (addr 0x0000 to 0x1fff) and write 1024 bytes to addr 0x1f00 (`flashpage_write_raw(0x1f00, SOME_DATA_PTR, 1024)`) -> this write will overflow!

To address this, I propose to do two things:
- add an additional assert in the `write_raw()` function: 
```c
assert((target_addr + len) < (CPU_FLASH_BASE + (FLASHPAGE_SIZE * FLASHPAGE_NUMOF)));
```
- add an additional note/sentence to the doxygen of the `flashpage_write_raw()` function stating that the user must make sure that the given data fits into the flash...